src/components/core/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,107 @@
+import {
+  ReactNode,
+  RefObject,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import styled from 'styled-components';
+import TextButton from '../Button/TextButton';
+import colors from '../colors';
+
+type Props = {
+  mainText: string;
+  children?: ReactNode;
+};
+
+const refContainsEventTarget = (ref: RefObject<HTMLDivElement>, event: any) => {
+  return ref.current && ref.current.contains(event.target);
+};
+
+function Dropdown({ mainText, children }: Props) {
+  const [isDropdownVisible, setIsDropdownVisible] = useState(false);
+
+  const handleClick = useCallback(() => {
+    setIsDropdownVisible(!isDropdownVisible);
+  }, [isDropdownVisible]);
+
+  const dropdownMenuRef = useRef<HTMLDivElement>(null);
+  const dropdownButtonRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    function handleDocumentClick(event: any) {
+      if (!isDropdownVisible) {
+        return;
+      }
+
+      const clickedOutsideMenu = !refContainsEventTarget(
+        dropdownMenuRef,
+        event
+      );
+      const clickedOnButton = refContainsEventTarget(dropdownButtonRef, event);
+
+      if (clickedOutsideMenu && !clickedOnButton) {
+        setIsDropdownVisible(false);
+      }
+    }
+
+    // we only need to add document listener if dropdown is open
+    if (isDropdownVisible) {
+      document.addEventListener('mousedown', handleDocumentClick);
+    }
+    return () => {
+      document.removeEventListener('mousedown', handleDocumentClick);
+    };
+  }, [dropdownMenuRef, isDropdownVisible]);
+
+  return (
+    <StyledDropdown>
+      <div
+        style={{ width: 'fit-content', alignSelf: 'flex-end' }}
```

Comment:
1. it's not ideal to pass an object in directly, it should be memoized with useMemo...
2. that said, could we just make this a styled component like `StyledDropdownButtonContainer` instead of having to use inline styles?

Reply:
1. ok I'll memoize
2. So these non styled-component divs are a workaround I found to set a ref for the button and box properly. Styled Components didn't like it when I set the ref on them directly, so I wrapped them in a div. [Though it seems like I should be able to set it directly](https://styled-components.com/docs/advanced#refs)
And once they were wrapped I had to add these styles to get the child elements to play nice being absolute positioned and what not..